h3py/h3core.pyx

```diff
@@ -0,0 +1,274 @@
+    mv = create_mv(ptr, n)
+
+    # todo: we can do this much more efficiently by using kRingDistances
+    if flag != 0:
+        raise H3ValueError("Couldn't run the fast version!")
+        # # fall back to `kRingDistances` and filter for appropriate distance. don't need to use sets
+        # n = h3c.maxKringSize(k)
+        # ptr = create_ptr(n)
+
+        # # hmmm. maybe use a cython array here instead
+        # cdef int[:n] distances
+        # distances[:] = 0
+
+
+        # h3c.kRingDistances(h, k, ptr, &distances[0])
+
+    return mv
+
+
+
+cpdef H3int[:] children(H3int h, res=None):
+    u.check_addr(h)
+
+    if res is None:
+        res = resolution(h) + 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    u.check_res(res)
+
+    n = h3c.maxH3ToChildrenSize(h, res)
+
+    ptr = create_ptr(n)
+    h3c.h3ToChildren(h, res, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+
+
+cpdef H3int[:] compact(const H3int[:] hu):
+    for h in hu:
+        u.check_addr(h)
+
+    ptr = create_ptr(len(hu))
+    flag = h3c.compact(&hu[0], ptr, len(hu))
+    mv = create_mv(ptr, len(hu))
+
+    if flag != 0:
+        raise H3ValueError('Could not compact set of hexagons!')
+
+    return mv
+
+# todo: https://stackoverflow.com/questions/50684977/cython-exception-type-for-a-function-returning-a-typed-memoryview
+# apparently, memoryviews are python objects, so we don't need to do the except clause
+cpdef H3int[:] uncompact(const H3int[:] hc, int res):
+    for h in hc:
+        u.check_addr(h)
+
+    N = h3c.maxUncompactSize(&hc[0], len(hc), res)
+
+    ptr = create_ptr(N)
+    flag = h3c.uncompact(
+        &hc[0], len(hc),
+           ptr, N,
+        res
+    )
+    mv = create_mv(ptr, N)
+
+    if flag != 0:
+        raise H3ValueError('Could not uncompact set of hexagons!')
+
+    return mv
+
+
+
+# weird return type here
+cpdef H3int num_hexagons(int resolution) except -1:
+    u.check_res(resolution)
+
+    return h3c.numHexagons(resolution)
+
+
+cpdef double hex_area(int resolution, unit='km') except -1:
+    u.check_res(resolution)
+
+    area = h3c.hexAreaKm2(resolution)
+
+    # todo: multiple units
+    convert = {
+        'km': 1.0,
+    }
+    area *= convert[unit]
+
+    return area
+
+cpdef double edge_length(int resolution, unit='km') except -1:
+    u.check_res(resolution)
+
+    length = h3c.edgeLengthKm(resolution)
+
+    # todo: multiple units
+    convert = {
+        'km': 1.0,
+    }
+    length *= convert[unit]
+
+    return length
+
+
+
+
+cpdef bool are_neighbors(H3int h1, H3int h2):
+    u.check_addr(h1)
+    u.check_addr(h2)
+
+    return h3c.h3IndexesAreNeighbors(h1, h2) == 1
+
+
+cpdef H3int edge(H3int origin, H3int destination) except 1:
+    u.check_addr(origin)
+    u.check_addr(destination)
+
+    if h3c.h3IndexesAreNeighbors(origin, destination) != 1:
+        raise H3ValueError('Hexes are not neighbors: {} and {}'.format(origin, destination))
```

Comment:
"Indexes are not neighbors" (one might be a pentagon)

Reply:
Agreed. Although I would love to see an "H3 Glossary" where all of these terms are defined unambiguously (see my comment below.)

Reply:
fixed